Sources/ScipioKit/Producer/FrameworkProducer.swift

```diff
@@ -74,94 +87,152 @@ struct FrameworkProducer {
+        for chunk in chunked {
+            await withTaskGroup(of: CacheSystem.CacheTarget?.self) { group in
+                for target in chunk {
+                    group.addTask {
+                        do {
+                            let restored = try await restore(target: target, cacheSystem: cacheSystem)
+                            return restored ? target : nil
+                        } catch {
+                            return nil
+                        }
+                    }
+                }
+                for await target in group.compactMap({ $0 }) {
+                    restored.insert(target)
+                }
+            }
+        }
+        return restored
+    }
+
+    // Return true if pre-built artifact is available (already existing or restored from cache)
+    private func restore(target: CacheSystem.CacheTarget, cacheSystem: CacheSystem) async throws -> Bool {
+        let product = target.buildProduct
         let frameworkName = product.frameworkName
         let outputPath = outputDir.appendingPathComponent(product.frameworkName)
         let exists = fileSystem.exists(outputPath.absolutePath)
 
-        let needToBuild: Bool
-        if exists, isConsumingCacheEnabled {
-            let isValidCache = await cacheSystem.existsValidCache(product: product)
-            if isValidCache {
+        if isConsumingCacheEnabled {
+            let isValidCache = await cacheSystem.existsValidCache(target: target)
+            if isValidCache && exists {
                 logger.info("✅ Valid \(product.target.name).xcframework is exists. Skip building.", metadata: .color(.green))
-                return
-            }
-            logger.warning("⚠️ Existing \(frameworkName) is outdated.", metadata: .color(.yellow))
-            logger.info("🗑️ Delete \(frameworkName)", metadata: .color(.red))
-            try fileSystem.removeFileTree(outputPath.absolutePath)
-            let restored = await cacheSystem.restoreCacheIfPossible(product: product)
-            needToBuild = !restored
-            if restored {
-                logger.info("✅ Restore \(frameworkName) from cache storage", metadata: .color(.green))
+                return true
+            } else {
+                if exists {
+                    logger.warning("⚠️ Existing \(frameworkName) is outdated.", metadata: .color(.yellow))
+                    logger.info("🗑️ Delete \(frameworkName)", metadata: .color(.red))
+                    try fileSystem.removeFileTree(outputPath.absolutePath)
+                }
+                let restoreResult = await cacheSystem.restoreCacheIfPossible(target: target)
+                switch restoreResult {
+                case .succeeded:
+                    logger.info("✅ Restore \(frameworkName) from cache storage", metadata: .color(.green))
+                    return true
+                case .failed:
+                    logger.warning("⚠️ Restoring \(frameworkName) is failed", metadata: .color(.yellow))
```

Comment:
Why not show detail error here?

Reply:
Added error feedback
https://github.com/giginet/Scipio/pull/54/commits/b2d6bc44d8a28bb32d1a567a0a90d06148b3388c